egress/src/filter.rs

```diff
@@ -0,0 +1,1382 @@
+                    label,
+                    fill_holes(
+                        temp_fromtime_priority,
+                        FromToTimes {
+                            from_time: first_time,
+                            to_time: last_time,
+                        },
+                    ),
+                );
+            }
+        }
+    }
+    let elapsed = now.elapsed();
+    eprintln!("Elapsed: {elapsed:.2?}");
+    eprintln!("len filter: {:?}", filter.len());
+    //println!("filter: {:?}", filter);
+    Ok(filter)
+}
+
+pub async fn get_filter(
+    conn: &PooledPgConn<'_>,
+    _from_time: DateTime<Utc>,
+    _to_time: DateTime<Utc>,
+    filter_label: FilterLabel,
+    filter_list: FilterTimeseriesTable,
+) -> Result<Option<Vec<FilterData>>, tokio_postgres::Error> {
+    // get the background filter list, and lookup this label
+    let filter = filter_list.get(&filter_label);
+    // create a structure to keep what is applicable
+    let mut applicable_ts: Vec<(i32, DateTime<Utc>, DateTime<Utc>)> = vec![];
+    // fill the structure
+    match filter {
+        Some(priorities) => {
+            // TODO: repace this with logic consistent with the new struct
+            applicable_ts.push((1, MIN_UTC, MAX_UTC));
+            _ = priorities.patches.first().unwrap().from_time;
+            _ = priorities.patches.first().unwrap().tsid;
+            _ = priorities.to_time;
+            //for prio in priorities {
+            //    // is this applicable?
+            //    let ft_t = cut_from_to_based_on_ts(
+            //        FromToTimes {
+            //            from_time: Some(from_time),
+            //            to_time: Some(to_time),
+            //        },
+            //        FromToTimes {
+            //            from_time: prio.from_time,
+            //            to_time: prio.to_time,
+            //        },
+            //    );
+            //    // have overlap
+            //    if let Some(times) = ft_t {
+            //        if let Some(tt) = times.to_time {
+            //            applicable_ts.push((prio.ts_id, times.from_time.unwrap(), tt));
+            //        } else {
+            //            // open ended (to time from request)
+            //            applicable_ts.push((prio.ts_id, times.from_time.unwrap(), to_time));
+            //        }
+            //    }
+            //}
+        }
+        None => return Ok(None), // no prioritized timeseries
+    }
+
+    // create sql to get right timeseries for filter
+    let mut sql = String::from("SELECT obsvalue, obstime FROM data");
+    let mut iter = applicable_ts.iter().peekable();
+    while let Some(x) = iter.next() {
+        let get_ts = format!(
+            "WHERE (timeseries = {} \
+                    AND obstime BETWEEN '{}' AND '{}')",
+            x.0, x.1, x.1,
+        );
+        sql += &get_ts;
+        if iter.peek().is_some() {
+            sql.push_str(" AND ");
+        }
+    }
```

Comment:
This query generation looks wrong. I don't think you can have multiple `WHERE` clauses in a query. I think you want a single `WHERE` at the start, then something like `OR (timeseries = {} AND obstime BETWEEN '{}' AND '{}')`.

Honestly I'm not a fan of interpolating like this though, as it means we lose protection against SQL injection, and can't use prepared queries. I would prefer an approach of doing N queries to get all the timeseries independently in parallel using a FuturesOrdered.

Reply:
Using futuresordered now